# Лабораторная работа №4  
## Задание 1. Поиск «жульничающего» кубика  

Мы загрузили массив 1cube.npy, где каждая строка — это отдельный кубик.  
Для каждого кубика рассчитали среднее значение и определили кубик с наибольшим отклонением от теоретического среднего.


In [1]:
import numpy as np 
from scipy.stats import ttest_1samp, chisquare
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
cubes = np.load("1cube.npy")
print(cubes.shape)
print(cubes)
for i in range(len(cubes)):
    print(f'Кубик:{i+1} {np.mean(cubes[i]):.2f}')
chet = np.argmax([np.mean(cube) for cube in cubes])+1
print(f'Жульничает кубик №:{chet}')

ImportError: DLL load failed while importing _pairwise_fast: Политика управления приложениями заблокировала этот файл.

## Задание 2. Оценка качества предсказаний и корреляция  

Мы загрузили массив 2pred.npy и разделили его на:
- Первые две строки (A) — истинные и предсказанные значения.
- Последние две строки (B) — данные для оценки корреляции.

Рассчитали метрики MAE, MSE, R² и коэффициент корреляции.


In [ ]:
pred = np.load('2pred.npy')
A = pred[:2, :]
y_true,y_pred = A
mae = np.mean(np.abs(y_true - y_pred))
mse = np.mean((y_true - y_pred) ** 2)
ss_res = np.sum((y_true - y_pred) ** 2)
ss_tot = np.sum((y_true - np.mean(y_true)) ** 2)
r2 = 1 - (ss_res / ss_tot)
mae1 = mean_absolute_error(y_true, y_pred)
mse1 = mean_squared_error(y_true, y_pred)
r21 = r2_score(y_true, y_pred)
print(f"С помощью numpy: MAE: {mae:.3f}, MSE: {mse:.3f}, R^2: {r2:.3f}")
print(f'С помощью scikit-learn: MAE:{mae1:.3f}, MSE:{mse1:.3f}, R^2:{r21:.3f}')
B = pred[2:, :]
x,y = B
corr_m = np.sum((x - np.mean(x)) * (y - np.mean(y)))
corr_n = np.sqrt(np.sum((x - np.mean(x)) ** 2) * np.sum((y - np.mean(y)) ** 2))
corr_a = corr_m/corr_n
corr = np.corrcoef(x,y)[0,1] #здесь мы проверяем 
print(f'Корреляция(через corrcoef): {corr:.3f}')
print(f'Корреляция(по формуле): {corr_a:.3f}')

С помощью numpy: MAE: 0.500, MSE: 0.375, R^2: 0.949
С помощью scikit-learn: MAE:0.500, MSE:0.375, R^2:0.949
Корреляция(через corrcoef): 0.983
Корреляция(по формуле): 0.983


##### По итогу результаты:
- Средняя абсолютная ошибка равно *MAE = 0.500*
- Среднеквадратичная ошибка равно *MSE = 0.375*
- Коэффициент детерминации равно *R^2 = 0.949*
- Корреляция через коэффициент Пирсона *r(x,y) = 0.983* 
- Через функцию corrcoef *corr = 0.983*

## Задание 3. Проверка кубиков статистическими тестами  

Для каждого кубика из Задания 1 провели:
- **χ²-тест**: проверяет, похоже ли распределение на равномерное.  
(Есть еще и второй способ с помощью **t-test**: проверяет, отличается ли среднее от теоретического значения 3.5.  )

In [ ]:
for i in range(len(cubes)):
    cube = cubes[i]
    # t_stat, p_value = ttest_1samp(cube, 3.5)
    # print(f"Кубик {i+1}:")
    # print(f"T-статистика: {t_stat:.3f}")
    # print(f"P-value: {p_value:.3f}")
    counts = np.bincount(cube, minlength=7)[1:]
    expected = [len(cube) / 6] * 6
    chi_stat, p_chi = chisquare(counts, expected)
    if p_chi < 0.05:
        print("Отвергаем H0: различия статистически значимы.")
    else:
        print("Не отвергаем H0: различия статистически незначимы.")
    print(f"  Chi²-статистика: {chi_stat:.3f}, P-value (χ²): {p_chi:.3f}\n")
    

NameError: name 'cubes' is not defined

Кубики с p-value < 0.05 (например, №2 и №4) демонстрируют значимые отклонения от равномерного распределения и могут считаться нечестными.
Остальные кубики (№1 и №3) можно считать честными, так как их распределения не отличаются от ожидаемого равномерного.